deepsubtree.go

```diff
@@ -119,10 +139,57 @@ func (dst *DeepSubTree) linkNode(node *Node) error {
 	return nil
 }
 
-// Set sets a key in the working tree with the given value.
-// Assumption: Node with given key already exists and is a leaf node.
-// Modified version of set taken from mutable_tree.go
+// Verifies the given operation matches up with the witness data.
+// Also, verifies and adds existence proofs related to the operation.
+func (dst *DeepSubTree) verifyOperationAndProofs(operation Operation, key []byte, value []byte) error {
+	if dst.witnessData == nil {
+		return errors.New("witness data in deep subtree is nil")
+	}
+	if dst.operationCounter >= len(dst.witnessData) {
+		return fmt.Errorf(
+			"operation counter in witness data: %d should be less than length of witness data: %d",
+			dst.operationCounter,
+			len(dst.witnessData),
+		)
+	}
+	traceOp := dst.witnessData[dst.operationCounter]
+	if traceOp.Operation != operation || !bytes.Equal(traceOp.Key, key) || !bytes.Equal(traceOp.Value, value) {
+		return fmt.Errorf(
+			"traceOp in witnessData: %s, %s, %s does not match up with write operation for key: %s, value: %s",
+			traceOp.Operation, string(traceOp.Key), string(traceOp.Value), string(key), string(value),
```

Comment:
This method is used also for reading.
```suggestion
			"traceOp in witnessData (%s, %s, %s) does not match up with executed operation (%s, %s, %s)",
			traceOp.Operation, string(traceOp.Key), string(traceOp.Value), 
			operation, string(key), string(value),
```